hypershift-operator/controllers/hostedcluster/createorupdate_annotation_enforcer_test.go

```diff
@@ -16,73 +16,115 @@ import (
+					Annotations: map[string]string{
+						hostedClusterAnnotation: annotationValue,
+					},
+				},
+			},
 		},
 		{
 			name: "Existing annotations are kept",
-			mutateFN: func(m *metav1.ObjectMeta) controllerutil.MutateFn {
+			obj: &corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "foo",
+					Namespace: "bar",
+				},
+			},
+			mutateFN: func(o crclient.Object) controllerutil.MutateFn {
 				return func() error {
-					m.Annotations = map[string]string{
+					o.SetAnnotations(map[string]string{
 						"foo": "bar",
-					}
+					})
 					return nil
 				}
 			},
-			verify: func(t *testing.T, cm *corev1.ConfigMap) {
-				if cm.Annotations["foo"] != "bar" {
-					t.Errorf("expected 'foo' annotation to be kept, but it's gone. Annotations: %+v", cm.Annotations)
-				}
+			expected: &corev1.ConfigMap{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "ConfigMap",
+					APIVersion: "v1",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "foo",
+					Namespace: "bar",
+					Annotations: map[string]string{
+						hostedClusterAnnotation: annotationValue,
+						"foo":                   "bar",
+					},
+				},
 			},
 		},
-	}
-
-	for _, tc := range testCases {
-		t.Run(tc.name, func(t *testing.T) {
-			obj := &corev1.ConfigMap{
+		{
+			name: "Do not annotate cluster scoped resources",
+			obj: &corev1.Namespace{
 				ObjectMeta: metav1.ObjectMeta{
 					Name: "foo",
 				},
-			}
-
-			// expectedObj is the original object + mutateFN + the annotation
-			expectedObj := obj.DeepCopy()
-			tc.mutateFN(&expectedObj.ObjectMeta)()
-			if expectedObj.Annotations == nil {
-				expectedObj.Annotations = map[string]string{}
-			}
-			expectedObj.Annotations[hostedClusterAnnotation] = "/foo"
+			},
+			mutateFN: func(_ crclient.Object) controllerutil.MutateFn {
+				return func() error { return nil }
+			},
+			expected: &corev1.Namespace{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "Namespace",
+					APIVersion: "v1",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "foo",
+				},
+			},
+		},
+	}
 
-			client := fake.NewClientBuilder().WithObjects(obj).Build()
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			client := fake.NewClientBuilder().WithObjects(tc.obj).Build()
 			providerFactory := createOrUpdateWithAnnotationFactory(upsert.New(false))
-			req := reconcile.Request{NamespacedName: types.NamespacedName{Namespace: obj.Namespace, Name: obj.Name}}
-
-			if _, err := providerFactory(req)(context.Background(), client, obj, tc.mutateFN(&obj.ObjectMeta)); err != nil {
+			if _, err := providerFactory(req)(context.Background(), client, tc.obj, tc.mutateFN(tc.obj)); err != nil {
 				t.Fatalf("CreateOrUpdate failed: %v", err)
 			}
 
-			actual := &corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Name: "foo"}}
-			if err := client.Get(context.Background(), crclient.ObjectKeyFromObject(actual), actual); err != nil {
+			var actualObj crclient.Object
+			switch tc.obj.(type) {
```

Comment:
Why do we do this? We type assert and then directly lose the type info again by assigning the asserted object to a variable of the interface type?

Reply:
:faceplam: Wow, that's special.  I must have been going down a different path at some point and left this here.  Nice catch!